### PR TITLE
refactor: Stop duplicating code between the low level API and the SDK

### PR DESCRIPTION
**Pull request by ddiakiteaneo**

### CODE REVIEW

File: ArmoniK.SDK.Client/src/SessionServiceImpl.cpp [hunk 1]
```diff
@@ -351,51 +301,40 @@ void SessionServiceImpl::CleanupTasks(const std::set<std::string> &task_ids) {
...
 
   while (tasks_iterator != task_ids.end()) {
-    armonik::api::grpc::v1::tasks::GetResultIdsResponse response;
     // List batch of results from the given tasks
-    channel_pool.WithChannel([&](const std::shared_ptr<::grpc::Channel> &channel) {
-      auto stub = armonik::api::grpc::v1::tasks::Tasks::NewStub(channel);
-
-      ::grpc::ClientContext context;
-      armonik::api::grpc::v1::tasks::GetResultIdsRequest request;
+    auto map_results = channel_pool.WithChannel([&](const std::shared_ptr<::grpc::Channel> &channel) {
```
Comment (fdenefaneo) [inline ArmoniK.SDK.Client/src/SessionServiceImpl.cpp:318]: `const auto &channel` ?

File: ArmoniK.SDK.Client/src/SessionServiceImpl.cpp [hunk 2]
```diff
@@ -351,51 +301,40 @@ void SessionServiceImpl::CleanupTasks(const std::set<std::string> &task_ids) {
...
+      std::vector<std::string> batched_ids;
       for (size_t i = 0; i < batch_size && tasks_iterator != task_ids.end(); ++i) {
-        *request.mutable_task_id()->Add() = *tasks_iterator;
+        batched_ids.push_back(std::move(*tasks_iterator));
```
Comment (fdenefaneo) [inline ArmoniK.SDK.Client/src/SessionServiceImpl.cpp:321]: The code could be more relevant using `emplace_back` here, don't you think @lemaitre-aneo ?

File: ArmoniK.SDK.Client/src/SessionServiceImpl.cpp [hunk 3]
```diff
@@ -289,16 +246,9 @@ void SessionServiceImpl::DropSession() {
...
-    ::grpc::ClientContext context;
-    armonik::api::grpc::v1::sessions::CancelSessionRequest request;
-    armonik::api::grpc::v1::sessions::CancelSessionResponse response;
-    *request.mutable_session_id() = session;
-    auto status =
-        armonik::api::grpc::v1::sessions::Sessions::NewStub(channel)->CancelSession(&context, request, &response);
-    if (!status.ok()) {
-      throw ArmoniK::Sdk::Common::ArmoniKSdkException("Unable to cancel session " + status.error_message());
-    }
+  auto reply = channel_pool.WithChannel([&](const std::shared_ptr<::grpc::Channel> &channel) {
```
[the remark this answers is not in the document]
Comment (ddiakiteaneo) [reply ArmoniK.SDK.Client/src/SessionServiceImpl.cpp:249]: It was the same signature before not `const auto &`. That being said, I'm not sure it will change anything in this case.

File: ArmoniK.SDK.Client/src/SessionServiceImpl.cpp [hunk 4]
```diff
@@ -351,51 +301,40 @@ void SessionServiceImpl::CleanupTasks(const std::set<std::string> &task_ids) {
   // Cancel the given tasks
   while (tasks_iterator != task_ids.end()) {
     channel_pool.WithChannel([&](const std::shared_ptr<::grpc::Channel> &channel) {
-      auto stub = armonik::api::grpc::v1::tasks::Tasks::NewStub(channel);
-      ::grpc::ClientContext context;
-      armonik::api::grpc::v1::tasks::CancelTasksRequest request;
+      std::vector<std::string> batched_ids;
       for (size_t i = 0; i < batch_size && tasks_iterator != task_ids.end(); ++i) {
-        *request.mutable_task_ids()->Add() = *tasks_iterator;
+        batched_ids.push_back(*tasks_iterator);
```
[the remark this answers is not in the document]
Comment (lemaitre-aneo) [reply ArmoniK.SDK.Client/src/SessionServiceImpl.cpp:306]: First is better intent:
- `push_back` to add an element that already exist (either copy or move)
- `emplace_back` to construct an element in situ

Second is actually safety, `push_back` takes only `const X&` and `X&&` and is guaranteed to call either the copy or the move constructor, nothing else. While `emplace_back` can take anything and call any constructor, even ones marked `explicit`. This might be what you want, but if you know that your element already exist, it is not what you want.

Third, it `push_back` plays nicer with argument deduction, and allows you to write code like that:
```cpp
struct Coord {
	int x;
	int y;
};

void add_coord(std::vector<Coord>& coords, int x, int y) {
	coords.push_back({x, y});
}
```

Also, I should highlight that for copy and move, the generated code is actually identical with both `push_back` and `emplace_back`

More information: https://stackoverflow.com/questions/10890653/why-would-i-ever-use-push-back-instead-of-emplace-back/36919571#36919571

File: ArmoniK.SDK.Client/src/SessionServiceImpl.cpp (hunk 2 above)
Comment (lemaitre-aneo) [reply ArmoniK.SDK.Client/src/SessionServiceImpl.cpp:321]: It would actually be harder to read to read because we could not be sure just seeing this line what happens to the vector (see my other comment).

tldr; use `push_back` for copy or move existing elements, use `emplace_back` to _construct_ a new object in situ

**Review by fdenefaneo — APPROVED** [review]

**Review by lemaitre-aneo — APPROVED** [review]
